plugins/cad/src/components/ResourceEditorDialog/components/PxeParametricEditor/nodes/widget/PxeRosterWidgetNode.tsx

```diff
@@ -0,0 +1,174 @@
+ * limitations under the License.
+ */
+
+import { Button, makeStyles } from '@material-ui/core';
+import AddIcon from '@material-ui/icons/Add';
+import DeleteIcon from '@material-ui/icons/Delete';
+import { get } from 'lodash';
+import React, { Fragment } from 'react';
+import { IconButton } from '../../../../../Controls';
+import { useEditorStyles } from '../../../FirstClassEditors/styles';
+import { PxeParametricEditorNodeProps } from '../../PxeParametricEditorNode';
+import {
+  PxeRosterWidgetEntry,
+  PxeValueType,
+} from '../../types/PxeConfiguration.types';
+import {
+  PxeResourceChangeRequest,
+  PxeValue,
+} from '../../types/PxeParametricEditor.types';
+import { createResourceChunkAfterChangeRequest } from '../../utils/createResourceChunkAfterChangeRequest';
+import { generateValueLabel } from '../../utils/generateLabelsForWidgets';
+import {
+  arrayWithItemRemoved,
+  arrayWithItemReplaced,
+} from '../../utils/general/immutableArrays';
+import { PxeParametricEditorNodeList } from '../../PxeParametricEditorNodeList';
+
+type RosterItemResourceChunk = {
+  readonly key: string;
+  readonly value: PxeValue;
+};
+
+type RosterValueType = PxeValueType.Object | PxeValueType.Array;
+
+// TODO Consider refactoring this component after Prettier settings update.
+export const PxeRosterWidgetNode: React.FC<PxeParametricEditorNodeProps> = ({
+  configurationEntry,
+  resourceChunk,
+  onResourceChangeRequest,
+}) => {
+  const { values, itemEntries } = configurationEntry as PxeRosterWidgetEntry;
+  const valueDescriptor = values[0];
+  const rosterValueType = valueDescriptor.type as RosterValueType;
+
+  const itemChunks = itemChunksFromValue(
+    get(resourceChunk, valueDescriptor.path),
+  );
+
+  const handleResourceChangeRequestForItem = (
+    itemIndex: number,
+    changeRequest: PxeResourceChangeRequest,
+  ) => {
+    const newItemChunk = createResourceChunkAfterChangeRequest(
+      itemChunks[itemIndex],
+      changeRequest,
+    ) as RosterItemResourceChunk;
+
+    onResourceChangeRequest({
+      valueDescriptor,
+      newValue: valueFromItemChunks(
+        arrayWithItemReplaced(itemChunks, itemIndex, newItemChunk),
+        rosterValueType,
+      ),
+    });
+  };
+
+  const handleItemAddition = () => {
+    const newItemChunk: RosterItemResourceChunk = {
+      key:
+        rosterValueType === PxeValueType.Array ? String(itemChunks.length) : '',
+      value: undefined,
+    };
+
+    onResourceChangeRequest({
+      valueDescriptor,
+      newValue: valueFromItemChunks(
+        [...itemChunks, newItemChunk],
+        rosterValueType,
+      ),
+    });
+  };
+
+  const handleItemDeletion = (itemIndex: number) => {
+    onResourceChangeRequest({
+      valueDescriptor,
+      newValue: valueFromItemChunks(
+        arrayWithItemRemoved(itemChunks, itemIndex),
+        rosterValueType,
+      ),
+    });
+  };
+
+  const editorClasses = useEditorStyles();
```

Comment:
Is it a hook?

Reply:
Yes, it's a part of Hook API of MUI System: https://mui.com/system/styles/basics/#hook-api